Fix socket-tests for proper SyncMessage usage

In my #87 branch I'm redoing a bunch of client code and testing infra, and it has surfaced some issues in the socket-tests.  There problems boil down to incorrect serialization and parsing of SyncMessages over the socket.

The proper way to take something from a socket's onmessage and turn it into a SyncMessage is this:

``` js
  var msg;
  try {
    // First parse the JSON back into a regular JS Object
    data = JSON.parse(data);
    msg = SyncMessage.parse(data);
  } catch(e) {
    // JSON.parse failed... give some error
  }

  // msg is now a proper SyncMessage *or* a SyncMessage error type if data was invalid.
```

Going the other way you simply take a `SyncMessage` and call its `stringify()` method, and you'll get proper JSON.

Another problem I see is that sometimes we aren't sending a Buffer where it's expected, and instead give a regular array:

```
  2) [Downstream Syncing with Websockets] PATCH should make the server respond with a RESPONSE SYNC SyncMessage after ending a downstream sync, and initiating an upstream sync:
     Uncaught AssertionError: [Rsync patch error: "TypeError: Object 72,101,108,108,111,32,119,111,114,108,100,33,10 has no method 'copy'"]: expected [TypeError: Object 72,101,108,108,111,32,119,111,114,108,100,33,10 has no method 'copy'] to not exist
```

This is another species of the deserialization code not working.  I think the right move here is to dump all the resolveToJSON stuff you have and make use of SyncMessage methods instead.

--- a/tests/lib/util.js
+++ b/tests/lib/util.js
@@ -105,12 +105,14 @@ function ensureFile(path, contents, jar, callback) {
 }
 
 
-function resolveToJSON(string) {
-  return SyncMessage.parse(JSON.parse(string));
-}
-
-function resolveFromJSON(obj) {
-  return obj.stringify();
+function toSyncMessage(string) {
+  try {
+    string = JSON.parse(string);
+    string = SyncMessage.parse(string);
+  } catch(e) {
+    expect(e, "[Error parsing a SyncMessage to JSON]").to.not.exist;
+  }
+  return string;
 }
 
 /**
@@ -299,7 +301,7 @@ function cleanupSockets(done) {
 function sendSyncMessage(socketPackage, syncMessage, callback) {
   var socket = socketPackage.socket;
   socketPackage.setMessage(callback);
-  socketPackage.socket.send(resolveFromJSON(syncMessage));
+  socketPackage.socket.send(syncMessage.stringify());
 }
 
 /**
@@ -321,7 +323,7 @@ var downstreamSyncSteps = {
       socketPackage.socket.once("message", socketPackage.onMessage);
 
       if (!customAssertions) {
-        message = resolveToJSON(message);
+        message = toSyncMessage(message);
 
         expect(message.type, "[Diffs error: \"" + (message.content && message.content.error) + "\"]").to.equal(SyncMessage.RESPONSE);
         expect(message.name).to.equal(SyncMessage.DIFFS);
@@ -357,7 +359,7 @@ var downstreamSyncSteps = {
       expect(err, "[Rsync patch error: \"" + err + "\"]").not.to.exist;
 
       var patchResponse = SyncMessage.response.patch;
-      socketPackage.socket.send(resolveFromJSON(patchResponse));
+      socketPackage.socket.send(patchResponse.stringify());
 
       cb();
     });
@@ -377,7 +379,7 @@ var upstreamSyncSteps = {
       socketPackage.socket.once("message", socketPackage.onMessage);
 
       if (!customAssertions) {
-        message = resolveToJSON(message);
+        message = toSyncMessage(message);
 
         expect(message).to.exist;
         expect(message.type).to.equal(SyncMessage.RESPONSE);
@@ -390,7 +392,7 @@ var upstreamSyncSteps = {
     });
 
     var requestSyncMessage = SyncMessage.request.sync;
-    socketPackage.socket.send(resolveFromJSON(requestSyncMessage));
+    socketPackage.socket.send(requestSyncMessage.stringify());
   },
   generateChecksums: function(socketPackage, data, customAssertions, cb) {
     if (!cb) {
@@ -403,7 +405,7 @@ var upstreamSyncSteps = {
       // Reattach original listener
       socketPackage.socket.once("message", socketPackage.onMessage);
 
-      message = resolveToJSON(message);
+      message = toSyncMessage(message);
       if (!customAssertions) {
         expect(message).to.exist;
         expect(message.type).to.equal(SyncMessage.REQUEST);
@@ -423,7 +425,7 @@ var upstreamSyncSteps = {
       path: data.path,
       srcList: data.srcList
     };
-    socketPackage.socket.send(resolveFromJSON(requestChksumMsg));
+    socketPackage.socket.send(requestChksumMsg.stringify());
   },
   patchServerFilesystem: function(socketPackage, data, fs, customAssertions, cb) {
     if (!cb) {
@@ -440,7 +442,7 @@ var upstreamSyncSteps = {
       socketPackage.socket.once("message", socketPackage.onMessage);
 
       if (!customAssertions) {
-        message = resolveToJSON(message);
+        message = toSyncMessage(message);
 
         expect(message.type, "[Diffs error: \"" + (message.content && message.content.error) + "\"]").to.equal(SyncMessage.RESPONSE);
         expect(message.name).to.equal(SyncMessage.PATCH);
@@ -487,7 +489,7 @@ function prepareDownstreamSync(finalStep, username, token, cb){
 
     var socketPackage = openSocket({
       onMessage: function(message) {
-        message = resolveToJSON(message);
+        message = toSyncMessage(message);
 
         expect(message).to.exist;
         expect(message.type).to.equal(SyncMessage.RESPONSE);
@@ -495,7 +497,7 @@ function prepareDownstreamSync(finalStep, username, token, cb){
         expect(message.content).to.be.null;
 
         socketPackage.socket.once("message", function(message) {
-          message = resolveToJSON(message);
+          message = toSyncMessage(message);
           expect(message).to.exist;
           expect(message.type).to.equal(SyncMessage.REQUEST);
           expect(message.name).to.equal(SyncMessage.CHKSUM);
@@ -866,8 +868,7 @@ module.exports = {
   socketURL: socketURL,
   username: uniqueUsername,
   createJar: jar,
-  resolveToJSON: resolveToJSON,
-  resolveFromJSON: resolveFromJSON,
+  toSyncMessage: toSyncMessage,
 
   // Connection helpers
   authenticate: authenticate,

--- a/tests/unit/socket-tests.js
+++ b/tests/unit/socket-tests.js
@@ -38,7 +38,8 @@ describe('[Downstream Syncing with Websockets]', function(){
 
        var socketPackage = util.openSocket(socketData, {
          onMessage: function(message) {
-           message = util.resolveToJSON(message);
+           message = util.toSyncMessage(message);
+
            expect(message).to.exist;
            expect(message.type).to.equal(SyncMessage.REQUEST);
            expect(message.name).to.equal(SyncMessage.CHKSUM);
@@ -83,7 +84,7 @@ describe('[Downstream Syncing with Websockets]', function(){
 
        var socketPackage = util.openSocket(socketData, {
          onMessage: function(message) {
-           message = util.resolveToJSON(message);
+           message = util.toSyncMessage(message);
            expect(message).to.exist;
            expect(message.type).to.equal(SyncMessage.REQUEST);
            expect(message.name).to.equal(SyncMessage.CHKSUM);
@@ -122,7 +123,7 @@ describe('[Downstream Syncing with Websockets]', function(){
        var socketPackage = util.openSocket(socketData, {
          onMessage: function(message) {
            // First, confirm server acknowledgment
-           message = util.resolveToJSON(message);
+           message = util.toSyncMessage(message);
            expect(message).to.exist;
            expect(message.type).to.equal(SyncMessage.REQUEST);
            expect(message.name).to.equal(SyncMessage.CHKSUM);
@@ -155,7 +156,7 @@ describe('[Downstream Syncing with Websockets]', function(){
 
        util.prepareDownstreamSync(result.username, result.token, function(syncData, fs, socketPackage) {
          util.downstreamSyncSteps.generateDiffs(socketPackage, syncData, fs, function(msg, cb) {
-           msg = util.resolveToJSON(msg);
+           msg = util.toSyncMessage(msg);
 
            expect(msg.type, "[Message type error: \"" + (msg.content && msg.content.error) +"\"]" ).to.equal(SyncMessage.RESPONSE);
            expect(msg.name).to.equal(SyncMessage.DIFFS);
@@ -177,10 +178,10 @@ describe('[Downstream Syncing with Websockets]', function(){
 
        util.prepareDownstreamSync('generateDiffs', result.username, result.token, function(syncData, fs, socketPackage) {
          util.downstreamSyncSteps.patchClientFilesystem(socketPackage, syncData, fs, function(msg, cb) {
-           msg = util.resolveToJSON(msg);
+           msg = util.toSyncMessage(msg);
            var startSyncMsg = SyncMessage.request.sync;
            util.sendSyncMessage(socketPackage, startSyncMsg, function(message){
-             message = util.resolveToJSON(message);
+             message = util.toSyncMessage(message);
 
              expect(message).to.exist;
              expect(message.type).to.equal(SyncMessage.RESPONSE);
@@ -202,7 +203,7 @@ describe('[Downstream Syncing with Websockets]', function(){
        util.prepareDownstreamSync(result.username, result.token, function(data, fs, socketPackage) {
          var startSyncMsg = SyncMessage.request.sync;
          util.sendSyncMessage(socketPackage, startSyncMsg, function(msg){
-           var msg = util.resolveToJSON(msg);
+           var msg = util.toSyncMessage(msg);
 
            expect(msg).to.exist;
            expect(msg.type).to.equal(SyncMessage.ERROR);
@@ -228,7 +229,7 @@ describe('[Upstream Syncing with Websockets]', function(){
         util.completeDownstreamSync(result.username, result.token, function(syncData, fs, socketPackage) {
           // Authorize a user, open a socket, authorize and complete a downstream sync
           util.upstreamSyncSteps.requestSync(socketPackage, function(message, cb) {
-            message = util.resolveToJSON(message);
+            message = util.toSyncMessage(message);
 
             expect(message).to.exist;
             expect(message.type, "[Error: \"" + (message && message.name) + "\"]" ).to.equal(SyncMessage.RESPONSE);
@@ -236,7 +237,7 @@ describe('[Upstream Syncing with Websockets]', function(){
 
             setTimeout(function() {
               util.upstreamSyncSteps.requestSync(socketPackage, function(message2, cb2) {
-                message2 = util.resolveToJSON(message2);
+                message2 = util.toSyncMessage(message2);
 
                 expect(message2).to.exist;
                 expect(message2.type, "[Error: \"" + (message2 && message2.name) + "\"]" ).to.equal(SyncMessage.RESPONSE);
@@ -267,7 +268,7 @@ describe('[Upstream Syncing with Websockets]', function(){
             util.completeDownstreamSync(result2.username, result2.token, function(syncData2, fs2, socketPackage2) {
               // Start an upstream sync with the first client of the user
               util.upstreamSyncSteps.requestSync(socketPackage, function(message, cb) {
-                message = util.resolveToJSON(message);
+                message = util.toSyncMessage(message);
 
                 expect(message).to.exist;
                 expect(message.type, "[Error: \"" + (message && message.name) + "\"]" ).to.equal(SyncMessage.RESPONSE);
@@ -277,7 +278,7 @@ describe('[Upstream Syncing with Websockets]', function(){
                 // get the second client of the same user to initiate an upstream sync
                 setTimeout(function() {
                   util.upstreamSyncSteps.requestSync(socketPackage2, function(message2, cb2) {
-                    message2 = util.resolveToJSON(message2);
+                    message2 = util.toSyncMessage(message2);
 
                     expect(message2).to.exist;
                     expect(message2.type, "[Error: \"" + (message2 && message2.name) + "\"]" ).to.equal(SyncMessage.RESPONSE);
@@ -310,7 +311,7 @@ describe('[Upstream Syncing with Websockets]', function(){
             util.completeDownstreamSync(result2.username, result2.token, function(syncData2, fs2, socketPackage2) {
               // Start an upstream sync with the first client of the user
               util.upstreamSyncSteps.requestSync(socketPackage, function(message, cb) {
-                message = util.resolveToJSON(message);
+                message = util.toSyncMessage(message);
 
                 expect(message).to.exist;
                 expect(message.type, "[Error: \"" + (message && message.name) + "\"]" ).to.equal(SyncMessage.RESPONSE);
@@ -319,7 +320,7 @@ describe('[Upstream Syncing with Websockets]', function(){
                 // After it's confirmed to be started, wait half the time before the lock should be broken and
                 // get the second client of the same user to initiate an upstream sync
                 util.upstreamSyncSteps.requestSync(socketPackage2, function(message2, cb2) {
-                  message2 = util.resolveToJSON(message2);
+                  message2 = util.toSyncMessage(message2);
 
                   expect(message2).to.exist;
                   expect(message2.type, "[Error: \"" + (message2 && message2.name) + "\"]" ).to.equal(SyncMessage.ERROR);

--- a/tests/unit/util-tests.js
+++ b/tests/unit/util-tests.js
@@ -215,7 +215,7 @@ describe('Test util.js', function(){
         var socketPackage = util.openSocket(socketData, {
           onMessage: function(message){
             // First, confirm server acknowledgment
-            message = util.resolveToJSON(message);
+            message = util.toSyncMessage(message);
             expect(message).to.exist;
             expect(message.type).to.equal(SyncMessage.REQUEST);
             expect(message.name).to.equal(SyncMessage.CHKSUM);
@@ -300,7 +300,7 @@ describe('Test util.js', function(){
 
         util.prepareDownstreamSync(username, result.token, function(syncData, fs, socketPackage) {
           util.downstreamSyncSteps.generateDiffs(socketPackage, syncData, fs, function(message, cb) {
-            message = util.resolveToJSON(message);
+            message = util.toSyncMessage(message);
 
             expect(message.type).to.equal(SyncMessage.RESPONSE);
             expect(message.name).to.equal(SyncMessage.DIFFS);
@@ -367,7 +367,7 @@ describe('Test util.js', function(){
           expect(fs instanceof FileSystem).to.equal.true;
 
           util.upstreamSyncSteps.requestSync(socketPackage, function(message, cb) {
-            message = util.resolveToJSON(message);
+            message = util.toSyncMessage(message);
 
             expect(message).to.exist;
             expect(message.type).to.equal(SyncMessage.RESPONSE);

